stdlib/hashmap.mc

```diff
@@ -0,0 +1,225 @@
+--
+-- A simple generic hashmap library.
+--
+-- TODO:
+--  - Resizing of buckets.
+--  - Conversion to and from association lists.
+--
+
+include "math.mc"
+include "option.mc"
+include "string.mc"
+
+let hashmapDefaultBucketCount = 100
+
+-- The base type of a HashMap object.
+--   k: Polymorphic key type
+--   v: Polymorphic value type
+type HashMap = {
+  buckets : [[{hash : Int, key : k, value : v}]],
+  nelems : Int,
+  eq : k -> k -> Bool,
+  hashfn : k -> Int
+}
+
+
+-- Returns an empty hashmap with a default number of buckets.
+--   eqArg: A function that specifies equalities between keys in the hashmap.
+--   hashfnArg: A function for computing the hash of a key value.
+let hashmapEmpty = lam eqArg. lam hashfnArg.
+  {buckets = makeSeq hashmapDefaultBucketCount [],
+   nelems = 0,
+   eq = eqArg,
+   hashfn = hashfnArg}
+
+-- An empty hashmap using strings as keys
+let hashmapStrEmpty =
+  -- An implementation of the djb2 hash function (http://www.cse.yorku.ca/~oz/hash.html)
+  recursive let djb2 = lam hash. lam s.
+    if null s then
+      hash
+    else
+      let newhash = addi (addi (muli hash 32) hash) (char2int (head s)) in
+      djb2 newhash (tail s)
+  in
+  hashmapEmpty eqstr (djb2 5381)
+
+-- Inserts a value into the hashmap
+--   key: The key to bind the value to.
+--   value: The value to be inserted.
+--   hm: The hashmap to insert the value into.
+-- [NOTE]
+--   The insertion uses a recursion which is not tail-recursive.
+let hashmapInsert = lam key. lam value. lam hm.
+  let hash = hm.hashfn key in
+  let idx = modi (absi hash) (length hm.buckets) in
+  let bucket = get hm.buckets idx in
+  let newEntry = {hash = hash, key = key, value = value} in
+  recursive let inserter = lam seq.
+    if null seq then
+      [newEntry]
+    else
+      let entry = head seq in
+      if neqi hash entry.hash then
+        cons entry (inserter (tail seq))
```

Comment:
Yes. This if statement is introduced under the assumption that the equality function is slow and that the hash function is good. Not sure about how this affects the cumulative execution time of the hashmap, but this will hopefully keep the variance down.